[CI] StatementParserTests testInvalidJoinPatterns failing

**Build Scans:**
- [elasticsearch-periodic #6249 / single-processor-node-tests](https://gradle-enterprise.elastic.co/s/bq2okuy2dhzu6)
- [elasticsearch-periodic-platform-support #6238 / sles-15_platform-support-unix](https://gradle-enterprise.elastic.co/s/tzpykhw2zaldi)
- [elasticsearch-periodic-platform-support #6238 / amazonlinux-2_platform-support-aws](https://gradle-enterprise.elastic.co/s/h2dhtt76hjqz6)
- [elasticsearch-periodic-platform-support #6238 / sles-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/rbsm7ulxsk2v2)
- [elasticsearch-periodic-platform-support #6238 / almalinux-8-aarch64_checkpart3_platform-support-arm](https://gradle-enterprise.elastic.co/s/itsrmdot4rjas)
- [elasticsearch-periodic-platform-support #6231 / sles-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/ssojw4bty4kzq)
- [elasticsearch-periodic-platform-support #6231 / rhel-7_platform-support-unix](https://gradle-enterprise.elastic.co/s/z5rl45bjul32q)
- [elasticsearch-periodic-platform-support #6231 / oraclelinux-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/ln4sbm2dllsvc)
- [elasticsearch-periodic-platform-support #6231 / rocky-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/22776jd4s6vpa)
- [elasticsearch-periodic-platform-support #6231 / windows-2022_checkpart3_platform-support-windows](https://gradle-enterprise.elastic.co/s/g3qnree4midxo)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:test" --tests "org.elasticsearch.xpack.esql.parser.StatementParserTests.testInvalidJoinPatterns" -Dtests.seed=D527E855ABF12EBA -Dtests.locale=en-IN -Dtests.timezone=Asia/Jayapura -Druntime.java=23
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.parser.StatementParserTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testInvalidJoinPatterns),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: a string containing "invalid index pattern [hiomnwvj:-8zxswukkrvd2q2!4spv,axjtofkzadfx:<oc2whlv-{now/M-1M{yyyy.MM.dd|+14:00}}>,tpdzrvlspo:-.32rr6v*syh*4], remote clusters are not supported in LOOKUP JOIN"
     but: was "line 1:87: mismatched input '+14:00}}>,"tpdzrvlspo":"""-.32rr6v*syh*4"""' expecting {'dissect', 'drop', 'enrich', 'eval', 'grok', 'keep', 'limit', 'mv_expand', 'rename', 'sort', 'stats', 'where', 'lookup', DEV_INLINESTATS, DEV_LOOKUP, DEV_JOIN_LEFT, DEV_JOIN_RIGHT}"
```

**Issue Reasons:**
- [8.x] 2 consecutive failures in step sles-12_platform-support-unix
- [8.x] 2 consecutive failures in step almalinux-8-aarch64_checkpart3_platform-support-arm
- [8.x] 14 failures in test testInvalidJoinPatterns (8.3% fail rate in 168 executions)
- [8.x] 2 failures in step sles-12_platform-support-unix (50.0% fail rate in 4 executions)
- [8.x] 2 failures in step almalinux-8-aarch64_checkpart3_platform-support-arm (66.7% fail rate in 3 executions)
- [8.x] 2 failures in step part-3 (5.4% fail rate in 37 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic (50.0% fail rate in 4 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic-platform-support (50.0% fail rate in 4 executions)
- [8.x] 2 failures in pipeline elasticsearch-pull-request (5.7% fail rate in 35 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -397,9 +397,6 @@ tests:
   issue: https://github.com/elastic/elasticsearch/issues/121966
 - class: org.elasticsearch.xpack.searchablesnapshots.hdfs.SecureHdfsSearchableSnapshotsIT
   issue: https://github.com/elastic/elasticsearch/issues/121967
-- class: org.elasticsearch.xpack.esql.parser.StatementParserTests
-  method: testInvalidJoinPatterns
-  issue: https://github.com/elastic/elasticsearch/issues/121968
 - class: org.elasticsearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT
   method: test {yaml=update/100_synthetic_source/stored text}
   issue: https://github.com/elastic/elasticsearch/issues/121991

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/IdentifierGenerator.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/IdentifierGenerator.java
@@ -79,13 +79,13 @@ public class IdentifierGenerator {
         }
 
         var pattern = maybeQuote(index.toString());
+        if (pattern.contains("|")) {
+            pattern = quote(unquoteIndexPattern(pattern));
+        }
+
         if (canAdd(Features.CROSS_CLUSTER, features)) {
             var cluster = maybeQuote(randomIdentifier());
             pattern = maybeQuote(cluster + ":" + pattern);
-        }
-
-        if (pattern.contains("|") && pattern.endsWith("\"") == false) {
-            pattern = quote(unquoteIndexPattern(pattern));
         }
 
         return pattern;

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/IdentifierGenerator.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/IdentifierGenerator.java
@@ -84,8 +84,8 @@ public class IdentifierGenerator {
             pattern = maybeQuote(cluster + ":" + pattern);
         }
 
-        if (pattern.contains("|") && pattern.contains("\"") == false) {
-            pattern = quote(pattern);
+        if (pattern.contains("|") && pattern.endsWith("\"") == false) {
+            pattern = quote(unquoteIndexPattern(pattern));
         }
 
         return pattern;

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/IdentifierGenerator.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/IdentifierGenerator.java
@@ -78,10 +78,11 @@ public class IdentifierGenerator {
             index.insert(0, "-");
         }
 
-        var pattern = maybeQuote(index.toString());
+        var pattern = index.toString();
         if (pattern.contains("|")) {
-            pattern = quote(unquoteIndexPattern(pattern));
+            pattern = quote(pattern);
         }
+        pattern = maybeQuote(pattern);
 
         if (canAdd(Features.CROSS_CLUSTER, features)) {
             var cluster = maybeQuote(randomIdentifier());

